File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/common/uicomponent/audioeffect/ChangeVoiceAdapter.java
package com.trtc.uikit.livekit.common.uicomponent.audioeffect;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.tencent.liteav.audio.TXAudioEffectManager.TXVoiceChangerType;
import com.trtc.uikit.livekit.R;
import com.trtc.uikit.livekit.common.core.LiveController;
import com.trtc.uikit.livekit.common.core.store.state.operation.MediaState;

import java.util.ArrayList;
import java.util.List;

public class ChangeVoiceAdapter extends RecyclerView.Adapter<ChangeVoiceAdapter.ViewHolder> {

    private final Context               mContext;
    private final LiveController        mLiveController;
    private final MediaState            mMediaState;
    private       List<ChangeVoiceItem> mData;
    private       int                   mSelectedPosition;

    public ChangeVoiceAdapter(Context context, LiveController liveController) {
        mContext = context;
        mLiveController = liveController;
        mMediaState = liveController.getMediaState();
        initData();
    }

    private void initData() {
        mData = new ArrayList<>();
        mData.add(new ChangeVoiceItem(mContext.getString(R.string.livekit_change_voice_none),
                R.drawable.livekit_select_none, TXVoiceChangerType.TXLiveVoiceChangerType_0));
        mData.add(new ChangeVoiceItem(mContext.getString(R.string.livekit_change_voice_child),
                R.drawable.livekit_change_voice_child, TXVoiceChangerType.TXLiveVoiceChangerType_1));
        mData.add(new ChangeVoiceItem(mContext.getString(R.string.livekit_change_voice_girl),
                R.drawable.livekit_change_voice_girl, TXVoiceChangerType.TXLiveVoiceChangerType_2));
        mData.add(new ChangeVoiceItem(mContext.getString(R.string.livekit_change_voice_uncle),
                R.drawable.livekit_change_voice_uncle, TXVoiceChangerType.TXLiveVoiceChangerType_3));
        mData.add(new ChangeVoiceItem(mContext.getString(R.string.livekit_change_voice_ethereal),
                R.drawable.livekit_change_voice_ethereal, TXVoiceChangerType.TXLiveVoiceChangerType_11));
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.livekit_anchor_settings_panel_item,
                parent, false);
        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        final ChangeVoiceItem changeVoiceItem = mData.get(position);
        holder.textTitle.setText(changeVoiceItem.title);
        holder.imageIcon.setImageResource(changeVoiceItem.icon);
        if (changeVoiceItem.type == mMediaState.audioInfo.changerType.get()) {
            holder.imageIcon.setBackgroundResource(R.drawable.livekit_settings_item_select_background);
            mSelectedPosition = mData.indexOf(changeVoiceItem);
        } else {
            holder.imageIcon.setBackgroundResource(R.drawable.livekit_settings_item_not_select_background);
        }
        holder.layoutRoot.setOnClickListener(view -> {
            int index = holder.getBindingAdapterPosition();
            if (index != RecyclerView.NO_POSITION) {
                mLiveController.getMediaController().setVoiceChangerType(changeVoiceItem.type);
                notifyItemChanged(index);
                notifyItemChanged(mSelectedPosition);
            }
        });
    }


    @Override
    public int getItemCount() {
        return mData.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        public LinearLayout layoutRoot;
        public TextView     textTitle;
        public ImageView    imageIcon;

        public ViewHolder(View itemView) {
            super(itemView);
            layoutRoot = itemView.findViewById(R.id.ll_root);
            textTitle = itemView.findViewById(R.id.tv_title);
            imageIcon = itemView.findViewById(R.id.iv_icon);
        }
    }

    public static class ChangeVoiceItem {
        public String             title;
        public int                icon;
        public TXVoiceChangerType type;

        public ChangeVoiceItem(String title, int icon, TXVoiceChangerType type) {
            this.title = title;
            this.icon = icon;
            this.type = type;
        }
    }
}

File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/component/audioeffect/view/ChangeVoiceAdapter.kt
package com.trtc.uikit.livekit.component.audioeffect.view

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.trtc.uikit.livekit.R
import io.trtc.tuikit.atomicxcore.api.AudioChangerType
import io.trtc.tuikit.atomicxcore.api.AudioEffectStore

class ChangeVoiceAdapter(
    private val context: Context,
    private val audioEffectStore: AudioEffectStore
) :
    RecyclerView.Adapter<ChangeVoiceAdapter.ViewHolder>() {

    private var data: MutableList<ChangeVoiceItem> = ArrayList()
    private var selectedPosition: Int = 0

    init {
        initData()
    }

    private fun initData() {
        data.add(
            ChangeVoiceItem(
                context.getString(R.string.common_change_voice_none),
                R.drawable.audio_effect_select_none, AudioChangerType.NONE
            )
        )
        data.add(
            ChangeVoiceItem(
                context.getString(R.string.common_change_voice_child),
                R.drawable.audio_effect_change_voice_child, AudioChangerType.CHILD
            )
        )
        data.add(
            ChangeVoiceItem(
                context.getString(R.string.common_change_voice_girl),
                R.drawable.audio_effect_change_voice_girl, AudioChangerType.LITTLE_GIRL
            )
        )
        data.add(
            ChangeVoiceItem(
                context.getString(R.string.common_change_voice_uncle),
                R.drawable.audio_effect_change_voice_uncle, AudioChangerType.MAN
            )
        )
        data.add(
            ChangeVoiceItem(
                context.getString(R.string.common_change_voice_ethereal),
                R.drawable.audio_effect_change_voice_ethereal, AudioChangerType.ETHEREAL
            )
        )
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val itemView = LayoutInflater.from(parent.context).inflate(
            R.layout.audio_effect_layout_change_voice_item,
            parent, false
        )
        return ViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val changeVoiceItem = data[position]
        holder.textTitle.text = changeVoiceItem.title
        holder.imageIcon.setImageResource(changeVoiceItem.icon)
        if (changeVoiceItem.type == audioEffectStore.audioEffectState.audioChangerType.value) {
            holder.imageIcon.setBackgroundResource(R.drawable.audio_effect_settings_item_select_background)
            selectedPosition = data.indexOf(changeVoiceItem)
        } else {
            holder.imageIcon.setBackgroundResource(R.drawable.audio_effect_settings_item_not_select_background)
        }
        holder.layoutRoot.setOnClickListener { view ->
            val index = holder.bindingAdapterPosition
            if (index != RecyclerView.NO_POSITION) {
                audioEffectStore.setAudioChangerType(changeVoiceItem.type)
                notifyItemChanged(index)
                notifyItemChanged(selectedPosition)
            }
        }
    }

    override fun getItemCount(): Int {
        return data.size
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val layoutRoot: LinearLayout = itemView.findViewById(R.id.ll_root)
        val textTitle: TextView = itemView.findViewById(R.id.tv_title)
        val imageIcon: ImageView = itemView.findViewById(R.id.iv_icon)
    }

    class ChangeVoiceItem(
        val title: String,
        val icon: Int,
        val type: AudioChangerType
    )
}